### ISSUE TITLE
bug: `stg pop / goto / push` fails when the operation leads to current directory being deleted.

### ISSUE THREAD

**Issue by NonLogicalDev** [stacked-git/stgit#179]
Very peculiar bug that happens if you are located in a directory added by the patch you are about to pop or push is deleted.

STG VERSION:
```
Stacked Git 1.5.dev2+g67fedd1
git version 2.33.0
Python 3.9.10
```

<details>
<summary>To Reproduce (test script)</summary>

```bash
#!/bin/bash

init_stg_repo() {
	local REPO_DIR=$1

	rm -rf $REPO_DIR
	mkdir -p $REPO_DIR

	( cd $REPO_DIR;
		git init
		echo "init" > INIT

		git add INIT
		git commit -m "INIT"

		stg init
	) > /dev/null
}

init_patch_stack() {
	local REPO_DIR=$1

	( cd $REPO_DIR;
		#------------------------------
		# Create a patch that modifies init.
		stg new mod-init -m "mod-init"

		echo INIT_MOD > INIT
		git add INIT
		stg refresh --index

		#------------------------------
		# Create a patch that creates DIR_A (and file inside)
		stg new add-a-dir -m "add-a-dir"

		mkdir -p DIR_A
		touch DIR_A/FILE
		git add DIR_A/FILE
		stg refresh --index

		#------------------------------
		# Create a patch that creates DIR_A/DIR_B (and file inside)
		stg new add-a-b-dir -m "add-a-b-dir"

		mkdir -p DIR_A/DIR_B
		touch DIR_A/DIR_B/FILE
		git add DIR_A/DIR_B/FILE
		stg refresh --index

		#------------------------------
		# Create a patch that deletes DIR_A/DIR_B (and file inside)
		stg new del-a-b-dir -m "del-a-b-dir"

		rm -rf DIR_A/DIR_B
		git add DIR_A
		stg refresh --index

		#------------------------------
		# Leave del-a-b-dir unapplied (from root folder)
		stg pop del-a-b-dir
	) > /dev/null 2>/dev/null
}

print_result() {
	echo "RESULT: ----------------------------------"
	if [[ $1 -ne 0 ]]; then
		echo ">> FAILED with code $1"
	else
		echo ">> PASSED"
	fi
}

print_repo_wd() {
	echo "::PWD:: [git-root]/$(git rev-parse --show-prefix)"
}

echo ">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>"
echo "BASELINE: git reset to mod-init"
echo "------------------------------------------"
init_stg_repo stg-delete-test
init_patch_stack stg-delete-test
( cd stg-delete-test; set -xe;
	print_repo_wd
	stg series -d
	cd DIR_A/DIR_B
	print_repo_wd

	#------------------------------
	# Try popping the patch
	git reset --hard $(stg id mod-init)

	echo "Where the heck am I?"
	pwd
)
print_result $?
echo "------------------------------------------"
echo


echo ">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>"
echo "TEST 1: stg pop"
echo "------------------------------------------"
init_stg_repo stg-delete-test
init_patch_stack stg-delete-test
( cd stg-delete-test; set -xe;
	print_repo_wd
	stg series -d
	cd DIR_A/DIR_B
	print_repo_wd

	#------------------------------
	# Try popping the patch
	stg pop
)
print_result $?
echo "------------------------------------------"
echo

echo ">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>"
echo "TEST 2: stg goto mod-init"
echo "------------------------------------------"
init_stg_repo stg-delete-test
init_patch_stack stg-delete-test
( cd stg-delete-test; set -xe;
	print_repo_wd
	stg series -d
	cd DIR_A/DIR_B
	print_repo_wd

	#------------------------------
	# Try going to earlier patch
	stg goto mod-init || exit $?
)
print_result $?
echo "------------------------------------------"
echo

echo ">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>"
echo "TEST 3: stg push"
echo "------------------------------------------"
init_stg_repo stg-delete-test
init_patch_stack stg-delete-test
( cd stg-delete-test; set -x;
	print_repo_wd
	stg series -d
	cd DIR_A/DIR_B
	print_repo_wd

	#------------------------------
	# Try pushing a patch that deletes current dir.
	stg goto mod-init
)
print_result $?
echo "------------------------------------------"
echo
```

</details>

<details>
<summary>To Reproduce (result)</summary>

```
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
BASELINE: git reset to mod-init
------------------------------------------
+ print_repo_wd
++ git rev-parse --show-prefix
+ echo '::PWD:: [git-root]/'
::PWD:: [git-root]/
+ stg series -d
+ mod-init    # mod-init
+ add-a-dir   # add-a-dir
> add-a-b-dir # add-a-b-dir
- del-a-b-dir # del-a-b-dir
+ cd DIR_A/DIR_B
+ print_repo_wd
++ git rev-parse --show-prefix
+ echo '::PWD:: [git-root]/DIR_A/DIR_B/'
::PWD:: [git-root]/DIR_A/DIR_B/
++ stg id mod-init
+ git reset --hard 2bbad60bae638c77cf09d81541e6b692d4b98581
HEAD is now at 2bbad60 mod-init
+ echo 'Where the heck am I?'
Where the heck am I?
+ pwd
/Users/oleg.utkin/Documents/Developer/Forks/stgit.testrepo/stg-delete-test/DIR_A/DIR_B
RESULT: ----------------------------------
>> PASSED
------------------------------------------

>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
TEST 1: stg pop
------------------------------------------
+ print_repo_wd
++ git rev-parse --show-prefix
+ echo '::PWD:: [git-root]/'
::PWD:: [git-root]/
+ stg series -d
+ mod-init    # mod-init
+ add-a-dir   # add-a-dir
> add-a-b-dir # add-a-b-dir
- del-a-b-dir # del-a-b-dir
+ cd DIR_A/DIR_B
+ print_repo_wd
++ git rev-parse --show-prefix
+ echo '::PWD:: [git-root]/DIR_A/DIR_B/'
::PWD:: [git-root]/DIR_A/DIR_B/
+ stg pop
Popped add-a-b-dir
fatal: Unable to read current working directory: No such file or directory
stg pop: git failed with code 128
RESULT: ----------------------------------
>> FAILED with code 2
------------------------------------------

>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
TEST 2: stg goto mod-init
------------------------------------------
+ print_repo_wd
++ git rev-parse --show-prefix
+ echo '::PWD:: [git-root]/'
::PWD:: [git-root]/
+ stg series -d
+ mod-init    # mod-init
+ add-a-dir   # add-a-dir
> add-a-b-dir # add-a-b-dir
- del-a-b-dir # del-a-b-dir
+ cd DIR_A/DIR_B
+ print_repo_wd
++ git rev-parse --show-prefix
+ echo '::PWD:: [git-root]/DIR_A/DIR_B/'
::PWD:: [git-root]/DIR_A/DIR_B/
+ stg goto mod-init
Popped add-a-b-dir -- add-a-dir
fatal: Unable to read current working directory: No such file or directory
stg goto: git failed with code 128
+ exit 2
RESULT: ----------------------------------
>> FAILED with code 2
------------------------------------------

>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
TEST 3: stg push
------------------------------------------
+ print_repo_wd
++ git rev-parse --show-prefix
+ echo '::PWD:: [git-root]/'
::PWD:: [git-root]/
+ stg series -d
+ mod-init    # mod-init
+ add-a-dir   # add-a-dir
> add-a-b-dir # add-a-b-dir
- del-a-b-dir # del-a-b-dir
+ cd DIR_A/DIR_B
+ print_repo_wd
++ git rev-parse --show-prefix
+ echo '::PWD:: [git-root]/DIR_A/DIR_B/'
::PWD:: [git-root]/DIR_A/DIR_B/
+ stg goto mod-init
Popped add-a-b-dir -- add-a-dir
fatal: Unable to read current working directory: No such file or directory
stg goto: git failed with code 128
RESULT: ----------------------------------
>> FAILED with code 2
------------------------------------------
```

</details>

It seems like Git itself handles this situation gracefully. It leaves the shell in a very confused state, in a non-existant directory, but it does not error out and completes operation.

My thoughts are that all STG's patch operations should ideally be conducted from the repo root directory (the `git rev-parse --show-toplevel` one) otherwise there is a risk of borking stg metadata by partially fulfilling a push or pop. (I also have a suspicion that fold family of commands might be vulnerable to the same bug)

**Comment by jpgrayson** [comment]
Thank you for reporting this and for the reproducing script, @NonLogicalDev! It is very much appreciated.

I have not yet dug into this, but just from the issue headline I believe this can happen. And I think you're on-target with your assessment.

I'll be looking at this in the near future.

**Comment by jpgrayson** [comment]
This is resolved in StGit 2.0.
